File: javascript/ssrf/productEnrichment.js
const express = require('express');
const http = require('http');
const axios = require('axios');

const router = express.Router();

// VULN 1: User-supplied source URL fetched via http.get() - SSRF
router.post('/products/import', (req, res) => {
    const { sourceUrl, categoryId } = req.body;
    // No URL validation - allows requests to internal network
    http.get(sourceUrl, (response) => {
        let data = '';
        response.on('data', (chunk) => { data += chunk; });
        response.on('end', () => {
            const products = JSON.parse(data);
            importProducts(categoryId, products);
            res.json({ imported: products.length, category: categoryId });
        });
    }).on('error', (err) => {
        res.status(500).json({ error: err.message });
    });
});

// VULN 2: User-supplied RSS feed URL fetched via axios.get() for blog/promotions - SSRF
router.post('/promotions/sync-feed', async (req, res) => {
    const { feedUrl, storeId } = req.body;
    // Allows attacker to make server request internal metadata at http://169.254.169.254
    const feedResponse = await axios.get(feedUrl, {
        headers: { 'Accept': 'application/rss+xml, application/xml' },
    });
    const feedContent = feedResponse.data;
    const promotions = parseFeed(feedContent);
    res.json({ storeId, syncedPromotions: promotions.length });
});

// VULN 3: User-controlled API base URL concatenated with path then fetched - SSRF
router.get('/products/enrich', async (req, res) => {
    const { apiBase, productId } = req.query;
    // apiBase is user-controlled - attacker can target internal services with path appended
    const enrichmentUrl = apiBase + '/products/' + productId;
    const enrichResponse = await axios.get(enrichmentUrl);
    res.json({ productId, enrichment: enrichResponse.data });
});

function importProducts(categoryId, products) {}
function parseFeed(content) { return []; }

module.exports = router;

File: ruby/ssrf/product_enrichment.rb
require 'sinatra'
require 'net/http'
require 'open-uri'
require 'json'

# VULN 1: User-supplied source URL fetched via Net::HTTP.get - SSRF allows probing internal network
post '/products/import' do
  source_url = params[:source_url]
  product_data = Net::HTTP.get(URI(source_url))
  products = JSON.parse(product_data)
  products.each { |p| Product.create(p) }
  { imported: products.length }.to_json
end

# VULN 2: User-supplied RSS feed URL fetched via open() - SSRF through open-uri to internal services
get '/products/rss_import' do
  feed_url = params[:feed_url]
  feed_content = open(feed_url).read
  items = parse_rss_feed(feed_content)
  { items: items }.to_json
end

# VULN 3: User-controlled API base URL string-concatenated then fetched - SSRF to attacker-controlled host
get '/products/sync' do
  api_base = params[:api_base]
  endpoint = api_base + '/v1/products'
  uri = URI(endpoint)
  response = Net::HTTP.get_response(uri)
  JSON.parse(response.body).to_json
end
